backend/svirt.pm

```diff
@@ -82,62 +86,93 @@ sub do_stop_vm {
             $self->run_ssh_cmd("$virsh undefine --snapshots-metadata $vmname");
         }
     }
+
+    # TODO: stream serial_terminal.txt with scp on the fly instead
+    $self->scp($self->serial_terminal_filename(), "serial_terminal.txt");
+
     return {};
 }
 
-# In list context returns pair ($stdout, $stderr). In void (and scalar)
-# context just logs stdout and stderr, returns nothing.
+# Log stdout and stderr and return them in a list (comped).
 sub get_ssh_output {
     my ($chan) = @_;
+    die 'No channel found' unless $chan;
 
-    my ($stdout, $errout) = ('', '');
+    my ($stdout, $stderr) = ('', '');
     while (!$chan->eof) {
         if (my ($o, $e) = $chan->read2) {
             $stdout .= $o;
-            $errout .= $e;
+            $stderr .= $e;
         }
     }
-    if (wantarray) {
-        return ($stdout, $errout);
-    }
-    else {
-        bmwqemu::diag "Command's stdout:\n$stdout" if length($stdout);
-        bmwqemu::diag "Command's stderr:\n$errout" if length($errout);
-    }
+    chomp($stdout, $stderr);
+    bmwqemu::diag("Command's stdout:\n$stdout") if length($stdout);
+    bmwqemu::diag("Command's stderr:\n$stderr") if length($stderr);
+    return ($stdout, $stderr);
 }
 
-# Sends command to libvirt host, logs stdout and stderr of the command,
-# returns exit status.
-#
-# Example:
-#   my $ret = $svirt->run_cmd("virsh snapshot-create-as snap1");
-#   die "snapshot creation failed" unless $ret == 0;
+=head2 run_cmd($cmd, %args);
+Runs command to libvirt host over SSH, logs stdout and stderr of the command.
+
+Returns either exit code itself or with stdout and stderr (both in blocking
+mode. C<$args{nonblock}> is for long running process, returns $ssh and $sock
+for handling the process.
+
+# Examples:
+    my $ret = $svirt->run_cmd("virsh snapshot-create-as snap1");
+    die "snapshot creation failed" unless $ret == 0;
+
+    my ($ret, $stdout, $stderr) = $svirt->run_cmd("grep -q '$marker' $log");
+    my ($ssh, $chan) = $svirt->run_cmd($cmd_full, nonblock => 1);
+=cut
 sub run_cmd {
-    my ($ssh, $cmd) = @_;
+    my ($ssh, $cmd, %args) = @_;
+    bmwqemu::log_call(@_);
 
     my $chan = $ssh->channel();
+    die 'No channel found' unless $chan;
+
     $chan->exec($cmd);
-    get_ssh_output($chan);
+    if ($args{nonblock}) {
+        bmwqemu::diag("Run command: '$cmd' (nonblock)");
+        return ($ssh, $chan);
+    }
+
+    my ($stdout, $errout) = get_ssh_output($chan);
     $chan->send_eof;
     my $ret = $chan->exit_status();
-    bmwqemu::diag "Command executed: $cmd, ret=$ret";
+    bmwqemu::diag("Command executed: '$cmd', ret=$ret");
     $chan->close();
+
+    if (wantarray) {
```

Comment:
Maybe you can prevent adding a new occasion of `wantarray` here.

Reply:
If you really insist, I can replace it with args, but is it really that dangerous that we cannot it? (nice feature, smaller code).

Reply:
Not really important, but alternatively you could use `?` instead of `if`.

Reply:
> I would prefer if you change it. Maybe just use `$args{wantarray}` instead. That would be passed explicitly on the caller-site.

Fixed.